docs/website/contents/for-developers/HardForkCombinatorRollbackMonotonicityProposal.md

```diff
@@ -0,0 +1,285 @@
+
+- DSR - Determination is Sensitive to Rollback - for chains C and D where C < D and len(C) <= len(C /\ D) + k, whether the current era of C can have a determined end on C but an undetermined or different end on D: Sensitive vs Insensitive
+
+- ZEE - Zero Epoch Eras - whether an era (in its entirety) can contain zero epochs: Allowed vs Disallowed
+
+- ZBE - Zero Block Eras - whether an era (in its entirety) can contain zero blocks: Allowed vs Disallowed
+
+- QSR - Queries are Sensitivity to Rollback - same as DSR but for HFC-related Node-To-Client queries instead of for validity of blocks: Sensitive vs Insensitive
+
+Note the following relations.
+
+- ZEE=Allowed implies ZBE=Allowed.
+- DSR=Insensitive implies QSR=Insensitive.
+
+## Table for the Current Cardano HFC
+
+The following table and associated notes classifies the current HFC implementation in terms of these dimensions.
+
+| DST | CCT | PIL | DSR | ZEE | ZBE | QSR |
+| --- | --- | --- | --- | --- | --- | --- |
+| I   | A   | P 1 | I   | 2   | 3   | I   |
+
+- (1) It's Possible, but only if there's a Chain Growth violation at a particular time during the epoch.
+
+- (2) The current HFC's behavior is inconsistent without making further assumptions.
+  Overall, ZEE is allowed, but it might cause some unexpected behavior.
+  In particular, forecasts always anticipate that there is at least one safe zone of slots in an era (which is not true for eras with zero epochs).
+  That mismatch could manifest as incorrect time-slot translations eg (within ledger rules execution, but not within Node-To-Client query reponses).
+
+- (3) ZBE is allowed when ticking a ledger state; the HFC will invoke multiple translation functions in sequence.
+  But cross-era forecasting cannot skip eras, so no headers after an epoch with zero blocks could be validated, and so blocks could not progatate in the net.
+  Moreover, today's node would not even mint blocks, since the leadership check (at least currently) relies on forecasting.
+
+## Table for an Particular Hypothetical Cardano HFC
+
+The following table and associated notes classifies an hypothetical HFC implementation in terms of these dimensions.
+Nicolas Frisby anticipates this implementation would be easier to understand, less likely to surprise users, more general (ie place fewer requirements on the block types), and not require excessive code changes.
+
+| DST | CCT | PIL | DSR | ZSE | ZBE | QSR |
+| --- | --- | --- | --- | --- | --- | --- |
+| S   | D   | I   | S 2 | D   | D 1 | I 3 |
+
+- (1) ZBE would be achievable for a prefix of eras when constructing the initial HFC ledger state (in the `ProtocolInfo` passed to the Consensus layer), but otherwise the HFC as a block type does not need to support ZBE at all.
+
+- (2) Esgen points out that perhaps DSR could/should still be Insensitive, and that wouldn't obviously lose too much simplicity.
```

Comment:
I wonder how that could even be possible. :thinking: Do you have an example?

Reply:
I think @amesgen would more easily be able to explain it than I can.

Reply:
Right now, the criterion for `singleEraTransition` to report the era transition for Shelley-based eras is:

 - Were there at least $k$ blocks after the voting deadline ($6k/f$ before an epoch boundary)?

We could change that to

 - Were there at least $k$ blocks after the voting deadline *or* was the last applied block within the last $3k/f$ slots of the epoch?

This choice would have DST=Sensitive, PIL=Impossible and DSR=Insensitive (apart from chain growth violations in the $3k/f$ slots after the voting deadline).

Reply:
> Were there at least $k$ blocks after the voting deadline ($6k/f$ before an epoch boundary)?

But don't we require 2k blocks after the voting deadline to actually transition to the next era?

Reply:
> This choice would have DST=Sensitive

Why? Because we would be counting slots? 

> PIL=Impossible

Because we would be counting slots, like the ledger does?

Reply:
> was the last applied block within the last $3k/f$ slots of the epoch?

and I don't get why this will ensure that a rollback won't cause a different era-transition decision (even if there are no CG violations).

(sorry for the many questions :pray: )